### ISSUE TITLE
Plotting for Model2D

### ISSUE THREAD

**Issue by dnjohnstone** [hyperspy/hyperspy#789]
As discussed in #536 plotting of 2-dimensional models is not yet supported, there is also a question regarding how best to actually do this from an interpretability perspective...

At present the general opinion was in favour of plotting the model and the data side-by-side (I was also thinking that automatically plotting the residual, may also be handy as people should be looking at that). Perhaps this should also come after the changes being discussed/made in #423 and associated.

**Comment by dnjohnstone** [comment]
This has become a little more pressing than I imagined because the absence of all functions related to plotting currently causes the multifit() method to fail for Model2D -- now that events and widgets are implemented though we should be in a good place to extend things like interactive range selection (which were the things putting me off doing this initially) to Model2D.

I will discuss this in person with @to266 in the next day or two but if anyone else is interested in this functionality and feels like adding their weight to this it would be good to know.

**Comment by dnjohnstone** [comment]
Further to that, looking at the code, my plan is roughly as follows:

1. Tidy up Model1D -- first by removing EELS specific stuff which really shouldn't be there and then by moving parts of the plotting code that are general to both 1D and 2D to BaseModel - looking at it I think quite a lot of the code that is now implemented with widgets and events will move to BaseModel.

2. Come up with implementations for the truly 2D plotting bits.

**Comment by francisco-dlp** [comment]
Can you briefly mention what did you have in mind when you wrote: "first by removing EELS specific stuff which really shouldn't be there"?

**Comment by dnjohnstone** [comment]
I thought we'd discussed previously that the low_loss and set_convolution_axis methods shouldn't be there? Or are they used for EDS model too?

**Comment by francisco-dlp** [comment]
You are right, that was written for EELS but I think that convolution can be useful for other 1D signals as well. We should probably rename the `low_loss` attribute though. That said, this feature needs to be refurbished and possibly implemented in `Model2D` too.

**Comment by vidartf** [comment]
Yeah, e.g. convolving with a point-spread function for 2D images could be useful.

**Comment by magnunor** [comment]
2D-convolution with a "probe" function + https://github.com/hyperspy/hyperspy/issues/707 and we're close to doing STEM-simulations :)

**Comment by dnjohnstone** [comment]
Ok - so conclusion is...

low-loss attribute --> function to convolve

set_convolution_axis method --> stays where it is and we make a 2D equivalent, at least the simplest solution to my mind is to do that for each case rather than a general version. That said perhaps scipy.ndimage.filters.convolve could be a way to generalise?

**Comment by francisco-dlp** [comment]
What about `Model1D.low_loss` -> `Model1D.convolved_with`?

The convolution must be refurbished. This is possibly one of the very few pieces of code in HyperSpy that hasn't changed at all since I wrote it for EELSLab almost 8 years ago! It works for what it does, but we should implement the feature in a more general, efficient and flexible way. For example, one should be able to convolve with a component or even a model, and he parameters of this convolution component/model should be parameters of the model. For example, if we know that an spectrum that we're fitting is blurred by a gaussian PSF but we don't know its FWHM, the FWHM of the PSF should be one of the parameters to optimize in the model.

**Comment by dnjohnstone** [comment]
I will have an initial PR together adding some basic plotting support for Model2D largely so that multifit will stop being broken... I plan to implement the following points:

1. Plotting will be in the form of 3 images next to each other in the order data-residual-model

2. I will not support the plotting of individual components overlaid on the model, because I think it will be too messy.

3. I will rename the attribute _model_line --> _model2plot in BaseModel and then assign it to _model_line and _model_image respectively in Model1D & Model2D... I think that this will mean that many of the "update" handling methods can be general and will therefore move to BaseModel

Does that all sound ok? I'm particularly unsure about whether point 3 will cause any problems @vidartf or @to266 perhaps that's more obvious to you?

**Comment by dnjohnstone** [comment]
@francisco-dlp I just remembered that you made some steps towards this with some data you were working on locally? How far did that get?

**Comment by francisco-dlp** [comment]
I just overlaid circles and points  as a representation of the centre and radius of a `Sphere2D` component. Two lines of code that I don't think that they'll be useful for this.

Some comments:

* Why plotting the residuals instead of the red-chisq? Also, maybe what to plot could be optional as, for big images, plotting too many things may cause inconvenience.
* I think that overlying representations of the component parameters (when feasible) on the signal plot is actually very useful. Now that we have proper widgets, that task of making them interactive is much simpler than it used to be.
* Generalising `_model_line` so that part of the code in `Model1D` can be moved to `BaseModel` sounds good. Maybe rename it to `_model_fig_repr` or something of this sort?

**Comment by dnjohnstone** [comment]
Thanks - agreed on your first and third comments.

About component parameters - in principle I also agree, in practice the issue is what to plot that won't look like a massive mess... 

Plotting the central point of each component for example would be better than nothing, probably applies to every component one can think of and is clearly possible - I can't think of any other simple parameter that would be general though?

**Comment by francisco-dlp** [comment]
For a simple Gaussian2D, one could plot e.g the FWHM, although I agree that making it interactive would be a little challenging for the general case (rotation + asymmetry), but possible. Anyway, this is certainly a separated issue unrelated to a first implementation of `Model2D.plot`.
